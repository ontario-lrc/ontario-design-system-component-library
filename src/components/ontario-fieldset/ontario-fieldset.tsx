import { Component, h, Prop, Watch } from '@stencil/core';

import { Fieldset } from './ontario-fieldset.interface';

import { CaptionType, CaptionTypes } from '../../utils/common/input-caption/input-caption.types';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { validatePropExists, validateValueAgainstArray } from '../../utils/validation/validation-functions';

@Component({
	tag: 'ontario-fieldset',
	styleUrl: 'ontario-fieldset.scss',
	shadow: true,
})
export class OntarioFieldset implements Fieldset {
	/**
	 * The text value used for the legend of the fieldset.
	 */
	@Prop({ mutable: true }) legend: string;

	/**
	 * The size of the fieldset legend. If no prop is passed, it will be `default`.
	 */
	@Prop({ mutable: true }) legendSize: CaptionType = 'default';

	/**
	 * Watch for changes to the legendSize prop.
	 * This is for validation purposes to make sure the legendSize matches one of the CaptionTypes.
	 */
	@Watch('legendSize')
	validateLegendSize() {
		const isValid = validateValueAgainstArray(this.legendSize, CaptionTypes);

		if (isValid) {
			return this.legendSize;
		} else {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(` legendSize ${this.legendSize} `)
				.addRegularText('for')
				.addMonospaceText(' <ontario-fieldset> ')
				.addRegularText('is not a valid type. A default size will be applied.')
				.printMessage();

			return 'default';
		}
	}

	/*
	 * Watch for changes in the `legend` prop for validation purposes.
	 */
	@Watch('legend')
	validateLegend() {
		this.validateLegendText(this.legend);
	}

	/**
	 * Print the legend warning message
	 */
	validateLegendText(newValue: string) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' legend ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-fieldset> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}
	componentWillLoad() {
		this.validateLegend();
		this.validateLegendSize();
	}

	private getClass() {
		switch (this.legendSize) {
			case 'large':
				return `ontario-fieldset__legend ontario-fieldset__legend--large`;
			case 'heading':
				return `ontario-fieldset__legend ontario-fieldset__legend--heading`;
			case 'default':
			default:
				return `ontario-fieldset__legend`;
		}
	}

	render() {
		return (
			<div class="ontario-form-group">
				<fieldset class="ontario-fieldset">
					<legend class={this.getClass()}>
						{this.legendSize === 'heading' ? <h1>{this.legend}</h1> : this.legend}
					</legend>
					<slot />
				</fieldset>
			</div>
		);
	}
}
