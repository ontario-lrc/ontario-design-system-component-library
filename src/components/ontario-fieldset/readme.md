# ontario-fieldset

Use the fieldset element to group related form elements.

## Usage guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca/components/detail/fieldsets.html) for current documentation guidance for fieldsets.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the fieldset component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Please see the [examples](#examples) below for how to configure the component.

## Examples

Example of a fieldset component.

```html
<ontario-fieldset
	legend="What is your delivery address?"
	level-size="heading"
>
    <!-- Other form web components can be added as children to group related form elements together -->
    <ontario-textarea ...></ontario-textarea>
    <ontario-radio-buttons ...></ontario-radio-buttons>
</ontario-aside>
```

<!-- Auto Generated Below -->

## Properties

| Property     | Attribute     | Description                                                                  | Type                                | Default     |
| ------------ | ------------- | ---------------------------------------------------------------------------- | ----------------------------------- | ----------- |
| `legend`     | `legend`      | The text value used for the legend of the fieldset.                          | `string`                            | `undefined` |
| `legendSize` | `legend-size` | The size of the fieldset legend. If no prop is passed, it will be `default`. | `"default" \| "heading" \| "large"` | `'default'` |

---

_Built with [StencilJS](https://stenciljs.com/)_
