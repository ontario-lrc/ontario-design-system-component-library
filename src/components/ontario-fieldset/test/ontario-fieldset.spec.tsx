import { newSpecPage } from '@stencil/core/testing';
import { OntarioFieldset } from '../ontario-fieldset';

describe('ontario-fieldset', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioFieldset],
				html: `<ontario-fieldset legend="What is your delivery address?" legend-size="heading"></ontario-aside>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render the default legend size if no legendSize prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioFieldset],
			html: `<ontario-fieldset legend="What is your delivery address?"></ontario-fieldset>`,
		});

		expect(page.root).toEqualHtml(`
			<ontario-fieldset legend="What is your delivery address?">
				<mock:shadow-root>
					<div class="ontario-form-group">
						<fieldset class="ontario-fieldset">
							<legend class="ontario-fieldset__legend">What is your delivery address?</legend>
							<slot></slot>
						</fieldset>
					</div>
				</mock:shadow-root>
			</ontario-fieldset>
		`);
	});
});
