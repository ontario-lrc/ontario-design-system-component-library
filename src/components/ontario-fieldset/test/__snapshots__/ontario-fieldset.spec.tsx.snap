// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-fieldset snapshot should render the expected html 1`] = `
<ontario-fieldset legend="What is your delivery address?" legend-size="heading">
  <mock:shadow-root>
    <div class="ontario-form-group">
      <fieldset class="ontario-fieldset">
        <legend class="ontario-fieldset__legend ontario-fieldset__legend--heading">
          <h1>
            What is your delivery address?
          </h1>
        </legend>
        <slot></slot>
      </fieldset>
    </div>
  </mock:shadow-root>
</ontario-fieldset>
`;
